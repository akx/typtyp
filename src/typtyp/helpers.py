from typing import Hashable, Iterable, TypeVar

THashable = TypeVar("THashable", bound=Hashable)


def unique_in_order(values: Iterable[THashable]) -> Iterable[THashable]:
    seen = set()
    for v in values:
        if v in seen:
            continue
        seen.add(v)
        yield v
