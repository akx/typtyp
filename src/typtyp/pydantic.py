from __future__ import annotations

from typing import Any, Iterable

from pydantic import BaseModel


def is_pydantic_model(t) -> bool:
    """
    Infallibly checks if a type is a Pydantic model. Never throws.
    """
    try:
        return issubclass(t, BaseModel)
    except Exception:
        return False


def get_pydantic_fields(tp) -> Iterable[tuple[str, Any]]:
    for name, f in sorted(tp.model_fields.items(), key=lambda item: item[0]):
        yield (name, f.annotation)
