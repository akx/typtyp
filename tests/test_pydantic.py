from typing import TYPE_CHECKING

import pytest

from tests.common import HairColor
from tests.helpers import check_with_tsc, typescript_from_types

if TYPE_CHECKING:
    import pydantic
else:
    pydantic = pytest.importorskip("pydantic")


class Head(pydantic.BaseModel):
    size: int
    hair_color: HairColor | None


class Person(pydantic.BaseModel):
    head: Head
    name: str


def test_pydantic(snapshot):
    code = typescript_from_types([Person, HairColor, Head])
    assert code == snapshot
    assert check_with_tsc(code)
