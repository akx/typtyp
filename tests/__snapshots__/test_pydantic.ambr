# serializer version: 1
# name: test_pydantic
  '''
  export interface Person {
  head: Head
  name: string
  }
  export const enum HairColor {
  RED = "red",
  BLACK = "black",
  PURPLE = "purple",
  BLUE = "blue",
  }
  export interface Head {
  hair_color: HairColor | null
  size: number
  }
  
  '''
# ---
